### PR TITLE
Add event clients to hive client

### PR DESCRIPTION
**Pull request by highker**
Add new event type to presto-hive so that hive page sink may post
information when a write finishes. For now, a dummy event client is
added to just print out a log line.

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveEventClient.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+import java.io.IOException;
+
+public class HiveEventClient
+        extends AbstractEventClient
+{
+    private static final Logger log = Logger.get(HiveEventClient.class);
+
+    @Inject
+    public HiveEventClient()
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveEventClient.java]: You shouldn't need an empty constructor

File: presto-hive/src/main/java/com/facebook/presto/hive/WriteCompletedEvent.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+    @EventField
+    public long getRows()
+    {
+        return rows;
+    }
+
+    public String toString()
+    {
+        return format("Query %s finishes write URI: %s, schema: %s, table: %s, partition: %s.",
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/WriteCompletedEvent.java]: This is not the correct formatting for a `toString`. Move this to the logging statement and remove this method.

Also, don't end in a period, since it gets mixed with the value

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveEventClient.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        extends AbstractEventClient
+{
+    private static final Logger log = Logger.get(HiveEventClient.class);
+
+    @Override
+    public <T> void postEvent(T event)
+            throws IOException
+    {
+        WriteCompletedEvent writeCompletedEvent = (WriteCompletedEvent) event;
+        log.info("Query " + writeCompletedEvent.getQueryId() + " completed");
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveEventClient.java]: Log some more info like the filename, type and anything else interesting

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java [hunk 4]
```diff
@@ -362,7 +377,44 @@ public HiveWriter createWriter(Page partitionColumns, int position, OptionalInt
                     conf,
                     typeManager);
         }
-        return new HiveWriter(hiveFileWriter, partitionName, isNew, fileNameWithExtension, write.toString(), target.toString());
+
+        String writerImplementation = hiveFileWriter.getClass().getName();
+
+        // TODO: the lambda takes an ugly parameter
+        Consumer<HiveWriter> onCommit = (HiveWriter hiveWriter) ->
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java]: Do you need the type on the lambda?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java [hunk 5]
```diff
@@ -362,7 +377,44 @@ public HiveWriter createWriter(Page partitionColumns, int position, OptionalInt
...
+        {
+            long size;
+            try {
+                size = hdfsEnvironment.getFileSystem(session.getUser(), path, conf).getFileStatus(path).getLen();
+            }
+            catch (IOException e) {
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java]: also catch runtime `catch (IOException | RuntimeException e)`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java [hunk 6]
```diff
@@ -362,7 +377,44 @@ public HiveWriter createWriter(Page partitionColumns, int position, OptionalInt
...
+                // Do not fail the query if file system is not available
+                size = -1;
+            }
+
+            Optional<Principal> principal = session.getIdentity().getPrincipal();
+            String principalName = principal.isPresent() ? principal.get().getName() : "EMPTY";
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java]: `principal.map(Principal::getName).orElse(null)`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java [hunk 7]
```diff
@@ -362,7 +377,44 @@ public HiveWriter createWriter(Page partitionColumns, int position, OptionalInt
...
+
+            WriteCompletedEvent event = new WriteCompletedEvent(
+                    session.getQueryId(),
+                    path.toString(),
+                    schemaName,
+                    tableName,
+                    partitionName.isPresent() ? partitionName.get() : "EMPTY",
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java]: `partitionName.orElse(null)`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java [hunk 8]
```diff
@@ -362,7 +377,44 @@ public HiveWriter createWriter(Page partitionColumns, int position, OptionalInt
...
+                    outputStorageFormat.getOutputFormat(),
+                    writerImplementation,
+                    nodeManager.getCurrentNode().getVersion(),
+                    nodeManager.getCurrentNode().getHttpUri().getHost(),
+                    principalName,
+                    nodeManager.getEnvironment(),
+                    null,   // TODO: session properties are not accessible through the interface; we might either (1) extent the interface, or (2) get the properties only we are interested in
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java]: We should add a method to get all properties (for this connector).  Let's do this in another PR.

File: presto-hive/src/main/java/com/facebook/presto/hive/WriteCompletedEvent.java [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+import javax.annotation.concurrent.Immutable;
+
+import java.util.Map;
+
+@Immutable
+@EventType("WriteCompletedEvent")
+public class WriteCompletedEvent
+{
+    private final String queryId;
+    private final String fileSystemUri;
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/WriteCompletedEvent.java]: maybe just call this `path`?  

cc @electrum or @raghavsethi

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java (hunk 4 above)
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java]: looks like so; btw the the lambda is kinda ugly given it needs to take a param so that the row number can be accessed. Any better suggestions?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveEventClient.java (hunk 3 above)
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveEventClient.java]: changed to something like this:
```2017-04-08 03:17:42 INFO Query 20170407_213242_00018_tdncf finishes write path: file:/var/folders/w0/25whm1sn4bxbfptdg11rx905h08qk2/T/1491600757329-0/DWRF.SNAPPY/test_8662763f-0597-460b-b265-47b137a7b04c, schema: test, table: test, partition: null, format: com.facebook.hive.orc.OrcOutputFormat
```

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveEventClient.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        extends AbstractEventClient
+{
+    private static final Logger log = Logger.get(HiveEventClient.class);
+
+    @Override
+    public <T> void postEvent(T event)
+            throws IOException
+    {
+        WriteCompletedEvent writeCompletedEvent = (WriteCompletedEvent) event;
+        log.info(format("Query %s finishes write path: %s, schema: %s, table: %s, partition: %s, format: %s",
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveEventClient.java]: Logger is already formatted

Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveEventClient.java]: Do we want this at info level?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java [hunk 11]
```diff
@@ -362,7 +377,40 @@ public HiveWriter createWriter(Page partitionColumns, int position, OptionalInt
                     conf,
                     typeManager);
         }
-        return new HiveWriter(hiveFileWriter, partitionName, isNew, fileNameWithExtension, write.toString(), target.toString());
+
+        String writerImplementation = hiveFileWriter.getClass().getName();
+
+        Consumer<HiveWriter> onCommit = (HiveWriter hiveWriter) ->
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java]: I don't think the lambda arg needs the type

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java [hunk 12]
```diff
@@ -362,7 +377,40 @@ public HiveWriter createWriter(Page partitionColumns, int position, OptionalInt
...
+        {
+            long size;
+            try {
+                size = hdfsEnvironment.getFileSystem(session.getUser(), path, conf).getFileStatus(path).getLen();
+            }
+            catch (IOException | RuntimeException e) {
+                // Do not fail the query if file system is not available
+                size = -1;
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java]: Let's make this optional rather than a bogus value

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java [hunk 13]
```diff
@@ -362,7 +377,40 @@ public HiveWriter createWriter(Page partitionColumns, int position, OptionalInt
...
+                    schemaName,
+                    tableName,
+                    partitionName.orElse(null),
+                    outputStorageFormat.getOutputFormat(),
+                    writerImplementation,
+                    nodeManager.getCurrentNode().getVersion(),
+                    nodeManager.getCurrentNode().getHttpUri().getHost(),
+                    session.getIdentity().getPrincipal().map(Principal::getName).orElse(null),
+                    nodeManager.getEnvironment(),
+                    null,   // TODO: extent the session interface so that it returns all properties
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java]: You can just loop over all the declared properties here

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveEventClient.java [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+        extends AbstractEventClient
+{
+    private static final Logger log = Logger.get(HiveEventClient.class);
+
+    @Override
+    public <T> void postEvent(T event)
+            throws IOException
+    {
+        WriteCompletedEvent writeCompletedEvent = (WriteCompletedEvent) event;
+        log.debug("Query %s finishes write path: %s, schema: %s, table: %s, partition: %s, format: %s",
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveEventClient.java]: I'm not sure I like `finishes`.. maybe use `wrote path` or `wrote file` instead?  @electrum ideas?

Also, add size if we have it.

File: presto-hive/src/main/java/com/facebook/presto/hive/WriteCompletedEvent.java [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+            String tableName,
+            String partitionName,
+            String storageFormat,
+            String writerImplementation,
+            String prestoVersion,
+            String serverAddress,
+            String principal,
+            String environment,
+            Map<String, String> sessionProperties,
+            long bytes,
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/WriteCompletedEvent.java]: This must be a boxed `Long`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveEventClient.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+        extends AbstractEventClient
+{
+    private static final Logger log = Logger.get(HiveEventClient.class);
+
+    @Override
+    public <T> void postEvent(T event)
+            throws IOException
+    {
+        WriteCompletedEvent writeCompletedEvent = (WriteCompletedEvent) event;
+        log.debug("Query %s wrote path to %s, schema: %s, table: %s, partition: %s, format: %s, size: %s",
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveEventClient.java]: Let's make this info level so it's logged by default

Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveEventClient.java]: Please update the args to match the new format string
```java
"File created: query: %s, schema: %s, table: %s, partition: '%s', format: %s, size: %s, path: %s"

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java [hunk 17]
```diff
@@ -180,6 +193,17 @@ public HiveWriterFactory(
 
         this.session = requireNonNull(session, "session is null");
 
+        this.nodeManager = requireNonNull(nodeManager, "nodeManager is null");
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java]: We don't need these extra newlines

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java [hunk 18]
```diff
@@ -362,7 +388,38 @@ public HiveWriter createWriter(Page partitionColumns, int position, OptionalInt
                     conf,
                     typeManager);
         }
-        return new HiveWriter(hiveFileWriter, partitionName, isNew, fileNameWithExtension, write.toString(), target.toString());
+
+        String writerImplementation = hiveFileWriter.getClass().getName();
+
+        Consumer<HiveWriter> onCommit = (hiveWriter) ->
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java]: Format like
```java
Consumer<HiveWriter> onCommit = hiveWriter -> {

File: presto-hive/src/main/java/com/facebook/presto/hive/WriteCompletedEvent.java [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+            String writerImplementation,
+            String prestoVersion,
+            String serverAddress,
+            String principal,
+            String environment,
+            Map<String, String> sessionProperties,
+            Long bytes,
+            long rows)
+    {
+        this.queryId = queryId;
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/WriteCompletedEvent.java]: Add null checks to all of these that should not be null

Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/WriteCompletedEvent.java]: And for the rest, add `@Nullable` (`javax.annotation`) to the getters and parameters

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveEventClient.java (hunk 16 above)
Comment (martint) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveEventClient.java]: More things in the server log? :/

We logging an event. Why do we even need this?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java [hunk 20]
```diff
@@ -95,6 +96,7 @@ public void configure(Binder binder)
         recordCursorProviderBinder.addBinding().to(ColumnarBinaryHiveRecordCursorProvider.class).in(Scopes.SINGLETON);
         recordCursorProviderBinder.addBinding().to(GenericHiveRecordCursorProvider.class).in(Scopes.SINGLETON);
 
+        binder.bind(EventClient.class).to(HiveEventClient.class).in(Scopes.SINGLETON);
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java]: Instead of doing this, install `EventModule`, then add a binding here:
```java
newSetBinder(binder, EventClient.class).addBinding().to(HiveEventClient.class).in(Scopes.SINGLETON);
```
The event module has a multi-event client that sends the events to all bound clients, allowing multiple event listeners

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java [hunk 21]
```diff
@@ -95,6 +96,7 @@ public void configure(Binder binder)
         recordCursorProviderBinder.addBinding().to(ColumnarBinaryHiveRecordCursorProvider.class).in(Scopes.SINGLETON);
         recordCursorProviderBinder.addBinding().to(GenericHiveRecordCursorProvider.class).in(Scopes.SINGLETON);
 
+        Multibinder.newSetBinder(binder, EventClient.class).addBinding().to(HiveEventClient.class).in(Scopes.SINGLETON);
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java]: Static import

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveEventClient.java [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+public class HiveEventClient
+        extends AbstractEventClient
+{
+    private static final Logger log = Logger.get(HiveEventClient.class);
+
+    @Override
+    public <T> void postEvent(T event)
+            throws IOException
+    {
+        WriteCompletedEvent writeCompletedEvent = (WriteCompletedEvent) event;
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveEventClient.java]: Add SuppressWarnings if this gives an unchecked warning

Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveEventClient.java]: Add an instanceof check here and ignore if it's the wrong event type

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java [hunk 23]
```diff
@@ -179,6 +191,12 @@ public HiveWriterFactory(
         }
 
         this.session = requireNonNull(session, "session is null");
+        this.nodeManager = requireNonNull(nodeManager, "nodeManager is null");
+        this.eventClient = requireNonNull(eventClient, "eventClient is null");
+
+        requireNonNull(hiveSessionProperties, "hiveSessionProperties is null");
+        this.sessionProperties = hiveSessionProperties.getSessionProperties().stream()
+                .collect(toImmutableMap(entry -> entry.getName(), entry -> session.getProperty(entry.getName(), entry.getJavaType()).toString()));
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java]: This can be `PropertyMetadata::getName`

Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java]: Wrap the args to `toImmutableMap `

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java [hunk 24]
```diff
@@ -362,7 +382,38 @@ public HiveWriter createWriter(Page partitionColumns, int position, OptionalInt
                     conf,
                     typeManager);
         }
-        return new HiveWriter(hiveFileWriter, partitionName, isNew, fileNameWithExtension, write.toString(), target.toString());
+
+        String writerImplementation = hiveFileWriter.getClass().getName();
+
+        Consumer<HiveWriter> onCommit = hiveWriter ->
+        {
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java]: This should be on previous line

**Comment by highker** [comment]
Merged
